### PR TITLE
use-structured-metadata

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/giantswarm/issues/30308

This PR attempts to move some high cardinality labels like resource and filename into structured metadata.
It also attempts to rename the node_name into node to match with the metric labels.

The code fixes both alloy and promtail configuration.

To be able to override the promtail pipelineStages field, I also had to move the promtail configuration to a template file

### CODE REVIEW

File: pkg/resource/logging-config/alloy/logging.alloy.template [hunk 1]
```diff
@@ -35,8 +35,8 @@ discovery.relabel "kubernetes_pods" {
 	}
 
 	rule {
-		source_labels = ["__meta_kubernetes_pod_node_name"]
-		target_label  = "node_name"
+		source_labels = ["__meta_kubernetes_pod_node"]
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/alloy/logging.alloy.template:38]:
```suggestion
		source_labels = ["__meta_kubernetes_pod_node_name"]
```

File: pkg/resource/logging-config/promtail-logging-config.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ type promtailExtraEnv struct {
 }
 
 type promtailConfigSnippets struct {
+	PipelineStages      string               `yaml:"pipelineStages" json:"pipelineStages"`
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/promtail-logging-config.go:38]: Using this https://github.com/grafana/helm-charts/blob/614fe8997c2a2cdd6e63b2c0e8816e76b875711f/charts/promtail/values.yaml#L504 for kubernetes pod scrape job

File: pkg/resource/logging-config/promtail-logging-config.go [hunk 3]
```diff
@@ -173,8 +183,12 @@ func GeneratePromtailLoggingConfig(lc loggedcluster.Interface) (string, error) {
         resource: resource
         namespace: namespace
       source: objectRef
+	- structured_metadata:
+	    resource:
+			filename:
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/promtail-logging-config.go:188]:
```suggestion
filename:
```

File: pkg/resource/logging-config/alloy/logging.alloy.template [hunk 4]
```diff
@@ -92,6 +92,20 @@ loki.process "kubernetes_pods" {
 	forward_to = [loki.write.default.receiver]
 
 	stage.cri { }
+
+	stage.structured_metadata {
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/alloy/logging.alloy.template:96]: https://grafana.com/docs/alloy/latest/reference/components/loki/loki.process/#stagestructured_metadata-block

File: pkg/resource/logging-config/alloy/logging.alloy.template (hunk 1 above)
Comment (QuentinBisson) [inline pkg/resource/logging-config/alloy/logging.alloy.template:38]:
```suggestion
		source_labels = ["__meta_kubernetes_pod_node_name"]
```

File: pkg/resource/logging-config/promtail-logging-config.go [hunk 5]
```diff
@@ -92,13 +93,25 @@ func GeneratePromtailLoggingConfig(lc loggedcluster.Interface) (string, error) {
 					Name: "NODENAME",
 					ValueFrom: promtailExtraEnvValuefrom{
 						FieldRef: promtailExtraEnvFieldref{
-							FieldPath: "spec.nodeName",
+							FieldPath: "regextraRelabelConfigexspec.nodeName",
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/promtail-logging-config.go:96]:
```suggestion
							FieldPath: "spec.nodeName",
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -25,6 +33,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Fix Alloy logs secret handling
 
+
```
Comment (QuentinBisson) [inline CHANGELOG.md:36]:
```suggestion
```

File: pkg/resource/logging-config/promtail-logging-config.go [hunk 7]
```diff
@@ -1,230 +1,42 @@
...
-type promtailExtraVolumeHostpath struct {
-	Path string `yaml:"path" json:"path"`
-}
-
-type promtailExtraVolumeMount struct {
-	Name      string `yaml:"name" json:"name"`
-	MountPath string `yaml:"mountPath" json:"mountPath"`
-	ReadOnly  bool   `yaml:"readOnly" json:"readOnly"`
-}
+var (
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/promtail-logging-config.go:15]: Replace go struct with template as pipeline stages would be of the type []map[string]interface{} but that cannot be marshalled so going with a template is definitely easier and it also reduces the code

File: pkg/resource/logging-config/promtail/logging-config.yaml.template [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    snippets:
+      addScrapeJobLabel: true
+      extraRelabelConfigs:
+        {{- if .IsWorkloadCluster }}
+        - action: keep
+          regex: kube-system|giantswarm
+          source_labels:
+          - __meta_kubernetes_namespace
+        {{- end }}
+        - source_labels: ['node_name']
```
Comment (QuentinBisson) [inline pkg/resource/logging-config/promtail/logging-config.yaml.template:19]: We relabel node_name into node and drop the node_name label

**Comment by QuentinBisson** [comment]
This PR has been tested with Promtail but will require some testing with alloy for logs to be sure the syntax is correct.

The interesting thing is that the structured metadata are still extracted as fields in the grafana ui but they are not added in the same place in the logql query (added after the stream labels but before the parsed labels)

image

**Comment by QuentinBisson** [comment]
Alloy was also tested on golem

File: pkg/resource/logging-config/alloy-logging-config.go [hunk 9]
```diff
@@ -24,7 +24,7 @@ var (
 
 func init() {
 	alloyLoggingTemplate = template.Must(template.New("logging.alloy").Funcs(sprig.FuncMap()).Parse(alloyLogging))
-	alloyLoggingConfigTemplate = template.Must(template.New("logging-config.yaml").Funcs(sprig.FuncMap()).Parse(alloyLoggingConfig))
+	alloyLoggingConfigTemplate = template.Must(template.New("logging-config.alloy").Funcs(sprig.FuncMap()).Parse(alloyLoggingConfig))
```
Comment (hervenicol) [inline pkg/resource/logging-config/alloy-logging-config.go:27]: Why the change of suffix? Does it have an impact on how alloy reads the file, or is it just a convention to name alloy config files `.alloy`?

File: pkg/resource/logging-config/alloy/logging.alloy.template [hunk 10]
```diff
@@ -228,10 +242,22 @@ loki.process "kubernetes_audit" {
...
+		values = [
+			"filename",
+		]
+	}
+
 	stage.labels {
 		values = {
-			namespace = null,
-			resource  = null,
+			namespace = "",
```
Comment (hervenicol) [inline pkg/resource/logging-config/alloy/logging.alloy.template:260]: What is the difference between `namespace = null` and `namespace = ""`?

File: pkg/resource/logging-config/alloy/logging.alloy.template [hunk 11]
```diff
@@ -228,10 +242,22 @@ loki.process "kubernetes_audit" {
...
+	stage.structured_metadata {
+		values = {
+			"resource" = "",
+			"filename" = "",
+		}
+	}
+
+	stage.label_drop {
+		values = [
+			"filename",
```
Comment (hervenicol) [inline pkg/resource/logging-config/alloy/logging.alloy.template:254]: Don't you need to also drop `resource` there?

File: pkg/resource/logging-config/alloy/logging.alloy.template (hunk 10 above)
Comment (QuentinBisson) [reply pkg/resource/logging-config/alloy/logging.alloy.template:260]: I'm not sure but in doubt, I followed the Doc https://grafana.com/docs/alloy/latest/reference/components/loki/loki.process/#stagelabels-block

File: pkg/resource/logging-config/alloy/logging.alloy.template (hunk 11 above)
Comment (QuentinBisson) [reply pkg/resource/logging-config/alloy/logging.alloy.template:254]: Nope because I did not make it a label in the stage.labels underneath

File: pkg/resource/logging-config/alloy-logging-config.go (hunk 9 above)
Comment (TheoBrigitte) [reply pkg/resource/logging-config/alloy-logging-config.go:27]: Can we then also rename the template file to align with this, also having the .yaml in the end would be more meaningful as this is a yaml file.

```suggestion
	alloyLoggingConfigTemplate = template.Must(template.New("logging-config.alloy.yaml").Funcs(sprig.FuncMap()).Parse(alloyLoggingConfig))
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Move high cardinality values into structured-metadata:
+  - kubernetes audit log `resource` label
+  - `filename` label
+  - log `output stream` label
+- Rename `node_name` label into `node` to match metric label.
```
Comment (TheoBrigitte) [inline CHANGELOG.md:16]: Is the node_name to node renaming really needed ? I would rather not doing it unless we have a really good reason for it.

**Review by TheoBrigitte — COMMENTED** [review]
Sounds fine, I did not review the full Promtail config I assumed it stayed mostly the same when you moved it to the template file.

File: CHANGELOG.md (hunk 12 above)
Comment (QuentinBisson) [reply CHANGELOG.md:16]: Yes we do it to align with the metric label as was decided in the refinement last week https://github.com/giantswarm/giantswarm/issues/30308#issuecomment-2355694791

File: pkg/resource/logging-config/alloy-logging-config.go (hunk 9 above)
Comment (QuentinBisson) [reply pkg/resource/logging-config/alloy-logging-config.go:27]: Also, the file is already in a folder named alloy so naming the file alloy/logging-config.alloy.yaml is weird right? Unless we put all templates (promtail and alloy) in the same folder

Comment (TheoBrigitte) [reply pkg/resource/logging-config/alloy-logging-config.go:27]: It makes it easier to go from the go code to the template file if they both have the same name.

File: CHANGELOG.md (hunk 12 above)
Comment (TheoBrigitte) [reply CHANGELOG.md:16]: Ok fine

File: pkg/resource/logging-config/alloy-logging-config.go (hunk 9 above)
Comment (QuentinBisson) [reply pkg/resource/logging-config/alloy-logging-config.go]: It's fixed in the new commit :)

**Review by TheoBrigitte — APPROVED** [review]
